File: lcd1602.c
#include <stdio.h>
#include <wiringPi.h>
#include <wiringPiI2C.h>
#include <string.h>

#define LCDAddr			0x27

int BLEN = 1;
int fd;
int fd2;

void write_word(int data){
	int temp = data;
	if ( BLEN == 1 )
		temp |= 0x08;
	else{
		temp &= 0xF7;
		printf("BLEN is not 1\n");
	}
	wiringPiI2CWrite(fd, temp);
}

void send_command(int comm){
	int buf;
	// Send bit7-4 firstly
	buf = comm & 0xF0;
	buf |= 0x04;			// RS = 0, RW = 0, EN = 1
	write_word(buf);
	delay(2);
	buf &= 0xFB;			// Make EN = 0
	write_word(buf);
	//printf("send_command buf bit7-4 is %d\n", buf);
	//delay(2000);

	// Send bit3-0 secondly
	buf = (comm & 0x0F) << 4;
	buf |= 0x04;			// RS = 0, RW = 0, EN = 1
	write_word(buf);
	delay(2);
	buf &= 0xFB;			// Make EN = 0
	write_word(buf);
	//printf("send_command buf bit3-0 is %d\n", buf);
}

void send_data(int data){
	int buf;
	// Send bit7-4 firstly
	buf = data & 0xF0;
	buf |= 0x05;			// RS = 1, RW = 0, EN = 1
	write_word(buf);
	delay(2);
	buf &= 0xFB;			// Make EN = 0
	write_word(buf);
	//printf("send_data buf bit7-4 is   %d\n", buf);
	//delay(2000);


	// Send bit3-0 secondly
	buf = (data & 0x0F) << 4;
	buf |= 0x05;			// RS = 1, RW = 0, EN = 1
	write_word(buf);
	delay(2);
	buf &= 0xFB;			// Make EN = 0
	write_word(buf);
	//printf("send_data buf bit3-0 is %d\n",buf);
}

void init(){
	send_command(0x33);	// Must initialize to 8-line mode at first
	delay(5);
	send_command(0x32);	// Then initialize to 4-line mode
	delay(5);
	send_command(0x28);	// 2 Lines & 5*7 dots
	delay(5);
	send_command(0x0C);	// Enable display without cursor
	delay(5);
	send_command(0x01);	// Clear Screen
	wiringPiI2CWrite(fd, 0x08);
}

void clear(){
	send_command(0x01);	//clear Screen
}

void write(int x, int y, char data[]){
	int addr, i;
	int tmp;
	if (x < 0)  x = 0;
	if (x > 15) x = 15;
	if (y < 0)  y = 0;
	if (y > 1)  y = 1;

	// Move cursor
	addr = 0x80 + 0x40 * y + x;
	send_command(addr);
	
	tmp = strlen(data);
	for (i = 0; i < tmp; i++){
		send_data(data[i]);
//		wiringPiI2CWrite(fd, data[i]);
		printf("Now   %c   is printed\n", data[i]);
	}
}

void ina219init(){
	u_int16_t config = 0;
	config |=(0 << 13 | 1 << 11 | 14 << 7 | 14 << 3 | 7);
	wiringPil2CWriteReg8(fd2, 0, config);
}
void calibrate(){
	u_int16_t digits;
	float min_lsb, swap;
	r_shunt = 0.1;
	min_lsb = 7 / 32767;

	current_lsb = min_lsb;
    digits=0;

	while( current_lsb > 0.0 ){//If zero there is something weird...
        if( (u_int16_t)current_lsb / 1){
        	current_lsb = (u_int16_t) current_lsb + 1;
        	current_lsb /= pow(10,digits);
        	break;
        }
        else{
        	digits++;
            current_lsb *= 10.0;
        }
    };
	swap = (0.04096)/(current_lsb*r_shunt);
    cal = (u_int16_t)swap;
    power_lsb = current_lsb * 20;
	wiringPil2CWriteReg8(fd2, 5, cal);
} 
int16_t shuntVoltageRaw(){
	u_int16_t ret;
	wiringPil2CWriteReg8(fd2, 1, 0);
	ret = wiringPil2CReadReg16(fd2, 1);
	return ret;
}

void main(){
	//lcd1602
	fd = wiringPiI2CSetup(LCDAddr);
	init();
	clear();
	//ina219
	fd2 = wiringPiI2CSetup(0x40);
	ina219init();
	calibrate();
	int v =shuntVoltageRaw();
	char str[100];
	int data = 100;
	//delay(2000);
	
	while (1)
	{
		write(0, 0, "A1 B2!");
		sprintf(str, "%d", v)
		write(1, 4, str);
	}
	
}

File: lcd1602.py
'''
    1602/1602A
    author：ZengXiaojie
    description：
        直接实例化 My1602 对象
        然后调用对象的 print_lcd(param1, param2, str) 方法
            ex：print_lcd(0, 0, 'Hello, world!')
            param1：从第几位开始，共16位（0-15），超出不显示，除非移动屏幕（其实1602一行有40位）
            param2：第几行，共有两行，0为第一行，1位第二行。
            str：要显示的数据
        每次打印字符时，建议适当清屏。
'''

import time
import smbus2
import sys

class My1602(object):
    BUS = smbus2.smbus2(1)
    LCD_ADDR = 0x27
    BLEN = 1

    # '''
    # 开关灯
    def turn_light(self, key):
        self.BLEN = key
        if key == 1:
            self.BUS.write_byte(self.LCD_ADDR, 0x08)
        else:
            self.BUS.write_byte(self.LCD_ADDR, 0x00)
    # '''

    def write_word(self, addr, data):
        temp = data
        if self.BLEN == 1:
            temp |= 0x08
        else:
            temp &= 0xF7
        self.BUS.write_byte(addr, temp)

	# 写命令
    def send_command(self, comm):
        # 发送7-4位数据
        buf = comm & 0xF0
        buf |= 0x04  # RS = 0, RW = 0, EN = 1
        self.write_word(self.LCD_ADDR, buf)
        time.sleep(0.002)
        buf &= 0xFB
        self.write_word(self.LCD_ADDR, buf)

        # 发送3-0位数据
        buf = (comm & 0x0F) << 4
        buf |= 0x04  # RS = 0, RW = 0, EN = 1
        self.write_word(self.LCD_ADDR, buf)
        time.sleep(0.002)
        buf &= 0xFB
        self.write_word(self.LCD_ADDR, buf)

	# 写数据
    def send_data(self, data):
        # 发送7-4位数据
        buf = data & 0xF0
        buf |= 0x05  # RS = 1, RW = 0, EN = 1
        self.write_word(self.LCD_ADDR, buf)
        time.sleep(0.002)
        buf &= 0xFB
        self.write_word(self.LCD_ADDR, buf)

        # 发送3-0位数据
        buf = (data & 0x0F) << 4
        buf |= 0x05  # RS = 1, RW = 0, EN = 1
        self.write_word(self.LCD_ADDR, buf)
        time.sleep(0.002)
        buf &= 0xFB
        self.write_word(self.LCD_ADDR, buf)
	
	# 初始化
    def __init__(self):
        try:
            self.send_command(0x33)
            time.sleep(0.005)
            self.send_command(0x32)
            time.sleep(0.005)
            self.send_command(0x28)
            time.sleep(0.005)
            self.send_command(0x0C)
            time.sleep(0.005)
            self.send_command(0x01)
            self.BUS.write_byte(self.LCD_ADDR, 0x08)
        except:
            return None
        else:
            return None

	# 清屏
    def clear_lcd(self):
        self.send_command(0x01)  # 清屏

	# 显示字符
    def print_lcd(self, x, y, str):
        if x < 0:
            x = 0
        if x > 15:
            x = 15
        if y < 0:
            y = 0
        if y > 1:
            y = 1

        addr = 0x80 + 0x40 * y + x
        self.send_command(addr)

        for chr in str:
            self.send_data(ord(chr))


if __name__ == '__main__':
    my1602 = My1602()
    # turn_light(0)  # 关闭背景灯光
    my1602.print_lcd(0, 0, 'Hello, world!')
    my1602.print_lcd(8, 1, 'by zxj')
    
